### ISSUE TITLE
geopmwrite is sent SIGKILL on Theta when run multi-node

### ISSUE THREAD

**Issue by bgeltz** [geopm/geopm#460]
The following script is being sent a SIGKILL intermittently on Theta:

```
#!/bin/bash
#COBALT -t 1:00:00
#COBALT -n 2
#COBALT -A intel
#COBALT --jobname debug
#COBALT --env JOBID=$jobid
#COBALT -u 2
#COBALT --run_users bgeltz:cantalup:bbaker:drguttma
#COBALT -q debug-flat-quad

module unload darshan
module load geopm

aprun -n2 -N1 -- geopmwrite POWER_PACKAGE package 0 200
```

The resulting log files show:
```
$ tail -n 1 296870.cobaltlog
Mon Dec 17 23:41:38 2018 +0000 (UTC) Info: task completed normally with an exit code of 137; initiating job cleanup and removal

$ cat 296870.error
[NID 00021] 2018-12-17 23:40:41 Apid 10036118: initiated application termination

$ cat 296870.output
Application 10036118 exit signals: Killed
Application 10036118 resources: utime ~1s, stime ~3s, Rss ~15584, inblocks ~32304, outblocks ~0
```

Reproduction of the issue is sporadic.  I see one of 3 scenarios when I try this:
1. Sometimes the job completes successfully with no issue.
2. Sometimes I see the "initiated application termination" message and the return code is 0.
3. Sometimes I see the above messages.

Runnng on a single node with `-n1 -N1` works without issue.  Requesting an interactive debug session of 2 nodes and using `-n1 -N1 -L <NODE_ID>` to force the command to run on one node or the other works without issue.  The only time I have a problem is when I try to run multi-node in a single command.

I suspect this is related to the failed jobs at scale on Theta, but have not yet been able to prove the connection.

**Comment by bgeltz** [comment]
I've instrumented the code path that geopmwrite uses for further debug.  It's here:
https://github.com/bgeltz/geopm/commit/307a2857e3c365fab8a7ec95cfbf065e13646458

**Comment by bgeltz** [comment]
Apparently this is not isolated to geopmwrite:

```
$ aprun -n2 -N1 -- geopmread MSR::PKG_POWER_LIMIT:PL1_POWER_LIMIT package 0
200
[NID 00001] 2019-01-08 22:42:17 Apid 10116960: initiated application termination
Application 10116960 exit signals: Killed
Application 10116960 resources: utime ~1s, stime ~3s, Rss ~17796, inblocks ~32308, outblocks ~0
```

I had to run this 10 times to see the ```Killed``` message.  I did consistently either:
```
[NID 00000] 2019-01-08 22:40:38 Apid 10116862: initiated application termination
```
or
```
[NID 00001] 2019-01-08 22:42:17 Apid 10116960: initiated application termination
```

**Comment by bgeltz** [comment]
This is starting to look like it's just the default behavior of aprun.  I put ```MPI_Init()``` and ```MPI_Finalize()``` in the beginning and end of the main() for ```geopmwrite```.  I was setting up the code so I could continue debug in DDT.  That code is here: https://github.com/bgeltz/geopm/commit/d34c0112574f692075a60a0514768058705893be

When I put the MPI calls in, I was on longer able to reproduce the problem.  It appears as though when MPI is not present to synchronize the ranks between nodes of a job, that ALPS will kill ranks that are slow to complete their task when at least one rank finishes on some node.

I'm doing some more testing now to confirm that hypothesis.

**Comment by bgeltz** [comment]
I've just root caused this to using the PMPI interface on Theta.  I've proved it with the following code:

https://gist.github.com/bgeltz/d376bad186cbb43cc365fb50db0fb2ff
 
In my testing just now, I landed on nid00003 and nid00004.  I ensured that darshan was unloaded before I compiled the code.  That was the only change I made from the default login environment.
 
If I run that code as-is, with the calls to MPI_Init() and MPI_Finalize() commented out of the main() but the functions present in the file, I am able to reproduce the SIGKILL.
 
If I simply comment out the MPI_Init() and MPI_Finalize() functions and change nothing else, the program completes successfully.
 
This seems to imply that any code that utilizes PMPI but does not actually utilize MPI for synchronizing job completion could suffer from this bug.
